models/article.go

```diff
@@ -184,21 +218,208 @@ func (db *DB) UnfavoriteArticle(userID int, articleID int) error {
+	case url.Values:
+		whereClause, args, skip = buildWhereClause("users.username", "author", value)
+	default:
+		whereClause, args, skip = buildWhereClause("users.username", value)
+	}
+
+	if !skip {
+		return db.Joins("JOIN users ON users.id = articles.user_id").
+			Where(whereClause, args)
+	}
+
+	return db
+}
+
+// FilterFavoritedBy filter articles favorited by user(s) username, value argument can be string|[]string|url.Values
+// If a url.Values provided, it must contains a query string 'favorited' param name.
+func (DB) FilterFavoritedBy(db *gorm.DB, value interface{}) *gorm.DB {
+	var whereClause string
+	var args interface{}
+	var skip = true
+
+	switch value.(type) {
+	case url.Values:
+		whereClause, args, skip = buildWhereClause("users.username", "favorited", value)
+	default:
+		whereClause, args, skip = buildWhereClause("users.username", value)
+	}
+
+	if !skip {
+		var ids []int
+		err := db.New().
+			Model(&User{}).
+			Where(whereClause, args).
+			Pluck("id", &ids).Error
+
+		if err != nil {
+			return db
+		}
+
+		return db.Joins("JOIN favorites ON favorites.article_id = articles.id").
+			Where("favorites.user_id IN (?)", ids)
+	}
+
+	return db
+}
+
+// Offset set the wanted offset (defaulf: 0) to an existing *gorm.DB instance.
+func (DB) Offset(db *gorm.DB, offset interface{}) *gorm.DB {
+	var offsetValue = offset
+
+	switch offset.(type) {
+	case url.Values:
+		queryParams := offset.(url.Values)
+		v := queryParams.Get("offset")
+
+		if v != "" {
+			intVal, err := strconv.Atoi(v)
+			if err != nil {
+				offsetValue = defaultOffset
+			} else {
+				offsetValue = intVal
+			}
+		} else {
+			offsetValue = defaultOffset
+		}
+	}
+
+	return db.Offset(offsetValue)
+}
+
+// Limit set the number of max articles to fetch (defaulf: 20) to an existing *gorm.DB instance.
+func (DB) Limit(db *gorm.DB, limit interface{}) *gorm.DB {
+	var limitValue = limit
+
+	switch limit.(type) {
+	case url.Values:
+		queryParams := limit.(url.Values)
+		v := queryParams.Get("limit")
+
+		if v != "" {
+			intVal, err := strconv.Atoi(v)
+			if err != nil {
+				limitValue = defaultLimit
+			} else {
+				limitValue = intVal
+			}
+		} else {
+			limitValue = defaultLimit
+		}
+	}
+
+	return db.Limit(limitValue)
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// Scopes															 		 //
+///////////////////////////////////////////////////////////////////////////////
 
 // Order articles by created_at DESC eager loading Tags and User
-func defaultScope(db *gorm.DB) *gorm.DB {
+func defaultArticleScope(db *gorm.DB) *gorm.DB {
 	return db.Order("articles.created_at desc").
 		Preload("Tags").
 		Preload("User")
 }
+
+///////////////////////////////////////////////////////////////////////////////
+// Private Methods															 //
+///////////////////////////////////////////////////////////////////////////////
+
+func buildWhereClause(field string, vars ...interface{}) (string, interface{}, bool) {
```

Comment:
Ex: in the FilterByTag()
I can invoke 
```go
// query is an instance of gorm.DB
FilterByTag(query, []string{"tag1", "tag2"})
// will call buildWhereClause("tags.name", string["tag1", "tag2"])
// whereClause, value, skip
// skip:  
//   true: don't apply the filter (ex: empty string, empty slice,...)
//   false: apply
-> tags.name IN (?), []string{"tag1", "tag2"}, false

FilterByTag(query, "tag1")
// will call buildWhereClause("tags.name", "tag1")
// whereClause, value, skip
-> tags.name = ?, "tag1", false

FilterByTag(query, url.Values{"tag": "tag1")
// will call buildWhereClause("tags.name", "tag", "tag1")
// whereClause, value, skip
-> tags.name = ?, "tag1", false
```
Example in handlers/article.
```go
r.ParseForm()
query = h.DB.Limit(query, r.Form) -> will lookup the 'limit' key in the map if exists
query = h.DB.Offset(query, r.Form) -> will lookup the 'offset' key in the map if exists
query = h.DB.FilterByTag(query, r.Form) -> will lookup the 'tag' key in the map if exists
query = h.DB.FilterAuthoredBy(query, r.Form) -> will lookup the 'author' key in the map if exists
query = h.DB.FilterFavoritedBy(query, r.Form) -> will lookup the 'favorited' key in the map if exists

err = query.Find(&articles).Error
```


so buildWhereClause() its just an helper function to construct a where clause for a kind of value type:
```go
// primitive value type int, string, slice
buildWhereClause(column, value)

// url.Values (map)
buildWhereClause(column, mapKey, map)
```
but I may change the signature of the different filter later to be more explicit like
```go
h.DB.FilterByTag(query,"tag1") -> query  tags.name = tag1
h.DB.FilterByTag(query, []int{1, 2, 3} -> query the tags.id IN (1,2,3)

// For a map (or url.Values) specify the key to lookup
h.DB.FilterByTag(query,"tag", request.Form) -> use the value of the 'tag' key in the map if exists

// or

// query, column, value
h.DB.FilterByTag(query,"tags.id", []int{1,2,3})
// query, column, mapKey, map
h.DB.FilterByTag(query,"tags.name", "tag", request.Form)
```